extensions-contrib/kubernetes-overlord-extensions/src/main/java/org/apache/druid/k8s/overlord/KubernetesTaskRunner.java

```diff
@@ -0,0 +1,571 @@
+                      if (job.getStatus().getSucceeded() != null) {
+                        completedPhase = new JobResponse(job, PeonPhase.SUCCEEDED);
+                      } else {
+                        completedPhase = new JobResponse(job, PeonPhase.FAILED);
+                      }
+                    } else {
+                      // the job is active lets monitor it
+                      completedPhase = monitorJob(task, k8sTaskId);
+                    }
+                  }
+                  TaskStatus status;
+                  if (PeonPhase.SUCCEEDED.equals(completedPhase.getPhase())) {
+                    status = TaskStatus.success(task.getId(), completedPhase.getLocation());
+                  } else {
+                    status = TaskStatus.failure(
+                        task.getId(),
+                        "Task failed %s: " + k8sTaskId
+                    );
+                  }
+                  if (completedPhase.getJobDuration().isPresent()) {
+                    status = status.withDuration(completedPhase.getJobDuration().get());
+                  }
+                  updateStatus(task, status);
+                  return status;
+                }
+                catch (Exception e) {
+                  log.error(e, "Error with task: %s", k8sTaskId);
+                  throw e;
+                }
+                finally {
+                  // publish task logs
+                  Path log = Files.createTempFile(task.getId(), "log");
+                  try {
+                    FileUtils.write(
+                        log.toFile(),
+                        client.getJobLogs(new K8sTaskId(task.getId())),
+                        StandardCharsets.UTF_8
+                    );
+                    taskLogPusher.pushTaskLog(task.getId(), log.toFile());
+                  }
+                  finally {
+                    Files.deleteIfExists(log);
+                  }
+                  client.cleanUpJob(new K8sTaskId(task.getId()));
+                  synchronized (tasks) {
+                    tasks.remove(task.getId());
+                  }
+                }
+              })
+          ));
+      return tasks.get(task.getId()).getResult();
+    }
+  }
+
+  JobResponse monitorJob(Task task, K8sTaskId k8sTaskId)
+  {
+    return monitorJob(client.getMainJobPod(k8sTaskId), task, k8sTaskId);
+  }
+
+  JobResponse monitorJob(Pod peonPod, Task task, K8sTaskId k8sTaskId)
+  {
+    if (peonPod == null) {
+      throw new ISE("Error in k8s launching peon pod for task %s", k8sTaskId);
+    }
+    log.info("monitorJob: Enable TLS Port is " + node.isEnableTlsPort());
+    log.info("Pod Host IP " + peonPod.getStatus().getPodIP());
+    TaskLocation location = TaskLocation.create(
+        peonPod.getStatus().getPodIP(),
+        DruidK8sConstants.PORT,
+        DruidK8sConstants.TLS_PORT,
+        node.isEnableTlsPort()
+    );
+    client.waitForProcessToStart(location, 10, TimeUnit.MINUTES);
```

Comment:
It seems that we dedicate one Overlord thread to each K8s task? That this thread is, essentially, the OL proxy for the K8s task? If so, perhaps we can describe this design in a comment.